File: 337. House Robber III.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def rob(self, root: TreeNode) -> int:
        def dfs(x: TreeNode):
            if not x: return [0, 0]
            l, r = dfs(x.left), dfs(x.right)
            return [l[1]+r[1]+x.val, max(l)+max(r)]
        return max(dfs(root))
            

File: 337. House Robber III.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    
    pair<int,int> dfs(TreeNode* x) {
        if (!x) return {0, 0};
        auto l=dfs(x->left), r=dfs(x->right);
        return {max(l.first,l.second)+max(r.first,r.second), x->val+l.first+r.first};
    }
    
    int rob(TreeNode* root) {
        auto pr = dfs(root);
        return max(pr.first, pr.second);
    }
};